File: app/src/main/java/org/checklist/comics/comicschecklist/util/AppRater.java
package org.checklist.comics.comicschecklist.util;

import org.checklist.comics.comicschecklist.R;

import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.content.SharedPreferences;
import android.net.Uri;
import android.support.v7.app.AlertDialog;

/**
 * Classe che lancia la finestra di dialogo per invitare l'utente a recensire l'app.
 * @author Francesco Bevilacqua
 */
public class AppRater {

    /**
     * Metodo che conta quante volte è stata lanciata l'app ed i giorni passati dall'installazione.
     */
    public static void app_launched (final Context mContext) {
        SharedPreferences prefs = mContext.getSharedPreferences (Constants.PREF_APP_RATER, 0);
        if (prefs.getBoolean (Constants.PREF_USER_DONT_RATE, false)) {
            return ;
        }

        SharedPreferences.Editor editor = prefs.edit();

        // Incremento il contatore dei lanci
        long launch_count = prefs.getLong(Constants.PREF_LAUNCH_COUNT, 0) + 1;
        editor.putLong(Constants.PREF_LAUNCH_COUNT, launch_count);

        // Prendo la data del primo lancio dell'applicazione
        Long date_firstLaunch = prefs.getLong(Constants.PREF_DATE_FIRST_LAUNCH, 0);
        if (date_firstLaunch == 0) {
            date_firstLaunch = System.currentTimeMillis();
            editor.putLong(Constants.PREF_DATE_FIRST_LAUNCH, date_firstLaunch);
        }

        // Aspetto almeno 7 giorni prima di aprire la finestra
        if (launch_count >= Constants.LAUNCHES_UNTIL_PROMPT) {
            if (System.currentTimeMillis() >= date_firstLaunch +
                    (Constants.DAYS_UNTIL_PROMPT * 24 * 60 * 60 * 1000)) {

                // Prepare dialog
                AlertDialog.Builder builder = new AlertDialog.Builder(mContext, R.style.AppCompatAlertDialogStyle);

                // Positive button: launch Google Play page
                builder.setPositiveButton(R.string.dialog_rate_button, new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialog, int id) {
                        dialog.dismiss();
                        mContext.startActivity(new Intent(Intent.ACTION_VIEW,
                                Uri.parse("market://details?id=" + mContext.getPackageName())));
                    }
                });

                // Negative button: avoid opening rate again
                builder.setNegativeButton(R.string.dialog_no_thanks_button, new DialogInterface.OnClickListener() {
                    public void onClick(DialogInterface dialog, int id) {
                        SharedPreferences prefs = mContext.getSharedPreferences(Constants.PREF_APP_RATER, 0);
                        final SharedPreferences.Editor editorPref = prefs.edit();
                        if (editorPref != null) {
                            editorPref.putBoolean(Constants.PREF_USER_DONT_RATE, true);
                            editorPref.apply();
                        }
                        dialog.dismiss();
                    }
                });

                // Neutral button: propose next time
                builder.setNeutralButton(R.string.dialog_late_button, new DialogInterface.OnClickListener() {
                    public void onClick(DialogInterface dialog, int id) {
                        dialog.dismiss();
                    }
                });

                // Add decoration
                builder.setTitle(R.string.app_name);
                builder.setMessage(R.string.dialog_rate_text);
                builder.show();
            }
        }

        editor.apply();
    }
}

File: app/src/main/java/org/checklist/comics/comicschecklist/util/AppRater.kt
package org.checklist.comics.comicschecklist.util

import org.checklist.comics.comicschecklist.R

import android.content.Context
import android.content.Intent
import android.net.Uri

import org.checklist.comics.comicschecklist.extensions.Loggable
import org.checklist.comics.comicschecklist.extensions.lazyLogger

import org.jetbrains.anko.alert
import org.jetbrains.anko.verbose

/**
 * Class which launch a dialog asking for rate the app.
 */
object AppRater: Loggable {

    private const val PREF_USER_NOT_RATING = "not_showing_again"
    private const val PREF_APP_RATER = "app_rater"
    private const val PREF_LAUNCH_COUNT = "launch_count"
    private const val PREF_DATE_FIRST_LAUNCH = "date_first_launch"
    private const val DAYS_UNTIL_PROMPT = 7
    private const val LAUNCHES_UNTIL_PROMPT = 7

    private val LOG by lazyLogger()

    /**
     * Method which is launched every time the app is opened.
     */
    fun appLaunched(context: Context) {
        val prefs = context.getSharedPreferences(PREF_APP_RATER, 0)
        if (prefs.getBoolean(PREF_USER_NOT_RATING, false)) {
            return
        }

        val editor = prefs.edit()

        // Increment counter of how many times user launched the app
        val launchCount = prefs.getLong(PREF_LAUNCH_COUNT, 0) + 1
        editor.putLong(PREF_LAUNCH_COUNT, launchCount)

        // Take date of first launch
        var dateFirstLaunch: Long? = prefs.getLong(PREF_DATE_FIRST_LAUNCH, 0)
        if (dateFirstLaunch == 0L) {
            dateFirstLaunch = System.currentTimeMillis()
            editor.putLong(PREF_DATE_FIRST_LAUNCH, dateFirstLaunch)
        }

        // Wait at least 7 days before opening rate window
        if (launchCount >= LAUNCHES_UNTIL_PROMPT) {
            if (System.currentTimeMillis() >= dateFirstLaunch!! + DAYS_UNTIL_PROMPT * 24 * 60 * 60 * 1000) {

                // Show dialog
                context.alert {
                    titleResource = R.string.app_name
                    messageResource = R.string.dialog_rate_text

                    positiveButton(R.string.dialog_rate_button) { dialog -> dialog.dismiss()
                        context.startActivity(Intent(Intent.ACTION_VIEW,
                                Uri.parse("market://details?id=" + context.packageName)))
                    }

                    negativeButton(R.string.dialog_no_thanks_button) { dialog ->
                        LOG.verbose("onClick - User refused to rate app..")
                        val editorPref = context.getSharedPreferences(PREF_APP_RATER, 0).edit()
                        if (editorPref != null) {
                            editorPref.putBoolean(PREF_USER_NOT_RATING, true)
                            editorPref.apply()
                        }

                        dialog.dismiss()
                    }

                    neutralPressed(R.string.dialog_late_button) { dialog -> dialog.dismiss() }
                }.show()
            }
        }

        editor.apply()
    }
}
